parlai/core/teachers.py

```diff
@@ -448,10 +448,20 @@ def observe(self, observation):
         """
         Process observation for metrics.
         """
+        self.metrics.clear_recent()
         if hasattr(self, 'lastY') and self.lastY is not None:
             self.metrics.evaluate_response(observation, self.lastY)
             self.custom_evaluation(self.last_act, self.lastY, observation)
             self.lastY = None
+        recent_metrics = self.metrics.report_recent()
+        if recent_metrics:
+            # for display purposes (display_model), take all accumulated
+            # metrics back into the original observation. This is an abuse of
+            # Messages being pointers
+            if 'metrics' in observation:
+                # override agent-level metrics if present
+                observation.pop('metrics')
+            observation['metrics'] = recent_metrics
```

Comment:
I do not understand why you pop the metric and then set it again. Could you explain this for me please?

Reply:
https://github.com/facebookresearch/ParlAI/blob/840843185f292669814fc04272de41af74af4bb6/parlai/core/torch_agent.py#L2017-L2027

We send "local metrics" from TorchAgent (see [here](https://parl.ai/docs/tutorial_metrics.html#local-metrics)) to the teacher in the observation using this same key. This lets us distinguish metrics provenance when we are using Multitasking with mixed batches: every teacher observes the metrics of its example.

We then accumulate those local metrics inside the teacher, along with any metrics the teacher computes in this method: 

https://github.com/facebookresearch/ParlAI/blob/840843185f292669814fc04272de41af74af4bb6/parlai/core/metrics.py#L773-L822

Now that we want to include a dictionary of both teacher AND user agent metrics (for display and logging purposes), we need to replace that old metrics dict from TorchAgent. We could "update" it, but just overwriting it seems better imho

Reply:
Thanks for the explanation.